### DISCUSSION TITLE
what is "滚动策略循环"?

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#18471] [General]
https://www.jianshu.com/p/168db3b69fac

一个 stream 由多个 ledger 组成；每个 ledger 根据基于时间或空间的滚动策略循环。

@sijie 

how do loop??

**Comment by startjava** [comment]
https://pulsar.apache.org/docs/next/concepts-architecture-overview/#managed-ledgers

A ledger can be deleted when all cursors have consumed the messages it contains. This allows for periodic rollover of ledgers.

“periodic rollover of ledgers”

what is rollover of ledgers ? how rollover??

ledgers not delete, from begin write data ??

**Comment by MarvinCai** [comment]
Rollover means delete old(consumed) ledgers and make room for new ledgers.
Ledger is append only and will not be writable once closed/fenced.
Ledger can only be deleted as a whole, so all entries on a ledger has to be consumed and acked by consumers(or auto acked by ttl).
Broker will scan ledger at background and if a ledger is fullly consumer(all entries are consumed) & (ledger exists more than configured retention period || total storage size for a topic cross max retion size), then it's marked for delete(rollover).
